mesonbuild/compilers/c.py

```diff
@@ -353,6 +353,26 @@ def __init__(self, exelist, version, for_machine: MachineChoice,
                            info, exe_wrap, **kwargs)
         MSVCCompiler.__init__(self, target)
 
+    def get_options(self):
+        opts = super().get_options()
+        c_stds = ['none', 'c89', 'c99', 'c11']
+        opts.update({
+            'std': coredata.UserComboOption(
+                'C language standard to use',
+                c_stds,
+                'none',
+            ),
+        })
+        return opts
+
+    def get_option_compile_args(self, options):
+        args = []
+        std = options['std']
+        # As of MVSC 16.7, /std:c11 is the only valid C standard option.
+        if std.value in {'c11'}:
+            args.append('/std:' + std.value)
```

Comment:
Why aren't we checking the version before passing this option? What happens if you pass this option to a version of MSVC that doesn't support it?

Reply:
FWIW, MSVC 16.7 is `cl` version `19.27.XXXXX`. 16.6 was `19.26.XXXXX`.

Reply:
I think we should have: 

1. A test that uses generics and uses `c_std=c11`, and is skipped on MSVC versions that don't support c11
2. A test that builds with `--werror` and uses `c_std=c11` unconditionally on some warning-free code, and is built on all MSVC versions

Reply:
For cv < 19.27, the /std:c11 flag is ignored by cl.exe with typical unknown flag warning.
Yes I agree there should be a test.
Separately, this same flag is coming in CMake 3.18.2.

Reply:
It would be more meaningful to the end user that they're doing something known not to be supported with older MSVC rather than emit compiler errors.

Reply:
> For cv < 19.27, the /std:c11 flag is ignored by cl.exe with typical unknown flag warning.

I think we should either error out that the c std is not supported on the current MSVC, or we should not pass `/std:c11` at all.

I think I prefer the latter since this is a preview feature and it's not clear what c11 features are supported out of the box with MSVC and which require the flag, and what features are not available even when the flag is passed.

Reply:
> It would be more meaningful to the end user that they're doing something known not to be supported with older MSVC rather than emit compiler errors.

Maybe we can emit a warning? I think we already do this for c++11 and c++14, f.ex..

Reply:
>  Separately, this same flag is coming in CMake 3.18.2.

I am happy to block 0.55.1 on this being fixed :)

Reply:
@scivision any updates here? Should we try to get this into 0.55.1?

Reply:
I may not have time for a while, maybe leave this out of 0.55.1.